### ISSUE TITLE
Cannot create bottle for tcl-tk with a newer brew than 3.3.11

### ISSUE THREAD

**Issue by imresteiner** [Homebrew/brew#12832]
### `brew config` output

```shell
> brew config
HOMEBREW_VERSION: 3.3.13-13-g62b82c0
ORIGIN: https://git.balabit/macos/homebrew.git
HEAD: 62b82c0792a3ad21584e6b1070214d2b1beaeded
Last commit: 12 seconds ago
Core tap ORIGIN: https://git.balabit/macos/homebrew-repo.git
Core tap HEAD: 88ac7fce91b3f1785d8d51767045ba5d8aa15253
Core tap last commit: 23 hours ago
Core tap branch: upgrade-formulas
HOMEBREW_PREFIX: /opt/bb-brew
HOMEBREW_CACHE: /opt/bb-brew/Caches
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_NO_AUTO_UPDATE: set
Homebrew Ruby: 2.6.8 => /opt/bb-brew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 11.0.3 build 1103
Git: 2.24.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.0.32.29
Xcode: 11.7
```

### `brew doctor` output

```shell
> brew doctor 
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo homebrew/core) checkout master

Warning: Your Xcode (11.7) is outdated.
Please update to Xcode 12.4 (or delete it).
Xcode can be updated from the App Store.
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I would like to update the formulas we are using to build our product. To release the binaries another git branch was created which is reseted to the homebrew-core repo's master.

### What happened (include all command output)?

`brew bottle --json tcl-tk` command failed.

output:
```
==> Determining tcl-tk bottle rebuild...
==> Bottling tcl-tk--8.6.12_1.catalina.bottle.tar.gz...
Error: new ID must be a String
Do not report this issue until you've run `brew update` and tried again.
/opt/bb-brew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/ruby-macho-3.0.0/lib/macho/macho_file.rb:318:in `change_dylib_id'
/opt/bb-brew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/ruby-macho-3.0.0/lib/macho/tools.rb:26:in `change_dylib_id'
/opt/bb-brew/Homebrew/Library/Homebrew/os/mac/keg.rb:10:in `change_dylib_id'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:26:in `block (2 levels) in relocate_dynamic_linkage'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/pathname.rb:341:in `ensure_writable'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:23:in `block in relocate_dynamic_linkage'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:22:in `each'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:22:in `relocate_dynamic_linkage'
/opt/bb-brew/Homebrew/Library/Homebrew/keg_relocate.rb:130:in `replace_placeholders_with_locations'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:501:in `block (2 levels) in bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/utils.rb:399:in `ignore_interrupts'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:499:in `ensure in block in bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:499:in `block in bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/keg.rb:334:in `block in lock'
/opt/bb-brew/Homebrew/Library/Homebrew/lock_file.rb:35:in `with_lock'
/opt/bb-brew/Homebrew/Library/Homebrew/keg.rb:330:in `lock'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:390:in `bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:105:in `block in bottle'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:104:in `each'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:104:in `bottle'
/opt/bb-brew/Homebrew/Library/Homebrew/brew.rb:110:in `<main>'
```
After a short investigation it seems an empty id is generated for the lib at Library/Homebrew/extend/os/mac/keg_relocate.rb:25
==> FILE: /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/tcltls1.7.22/tcltls.dylib
==> FILE.DYLIB_ID shared-tcltls.dylib
==> ID:

With version 3.3.11 brew can build the bottle successfully. With a newer version I could reproduce the issue.

### What did you expect to happen?

The json config is generated to create a bottle afterwards.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
> brew install --build-bottle tcl-tk
> brew bottle --json tcl-tk
> brew bottle --merge tcl-tk*.json --write
```

**Comment by imresteiner** [comment]
Finally finished with the git bisect.
It seems this commit brought in the issue:
                                                                                                        
568bc94f3ee66bb62a31ae9d9c7f15de8c60cc92 is the first bad commit
commit 568bc94f3ee66bb62a31ae9d9c7f15de8c60cc92
Author: Carlo Cabrera <30379873+carlocab@users.noreply.github.com>
Date:   Wed Jan 19 00:34:39 2022 +0800

    os/mac/keg: use `MachOFile#delete_rpath` instead of `MachO::Tools`
    
    This will allow us to avoid keeping track of the number of `RPATH`s
    while trying to delete duplicates.
    
    See discussion at #12745.

 Library/Homebrew/extend/os/mac/keg_relocate.rb | 10 +---------
 Library/Homebrew/os/mac/keg.rb                 |  2 +-
 Library/Homebrew/os/mac/mach.rb                |  7 +++++++
 3 files changed, 9 insertions(+), 10 deletions(-)

**Comment by carlocab** [comment]
Thanks. That's the only closely related change recently, so it makes sense. On the other hand, I don't see how that affects what currently looks like unrelated code...

**Comment by carlocab** [comment]
Can you show the output of `brew bottle --verbose --debug --json tcl-tk`?

**Comment by imresteiner** [comment]
> brew bottle --json --verbose --debug tcl-tk
/opt/bb-brew/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /opt/bb-brew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/tcl-tk.rb
==> Determining tcl-tk bottle rebuild...
/opt/bb-brew/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /opt/bb-brew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/tcl-tk.rb
/opt/bb-brew/Homebrew/Library/Homebrew/shims/shared/git --version
==> Bottling tcl-tk--8.6.12_1.catalina.bottle.tar.gz...
==> Changing install name in /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/bin/tclsh8.6
  from /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
    to @@HOMEBREW_CELLAR@@/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
==> Changing install name in /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/bin/wish8.6
  from /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtk8.6.dylib
    to @@HOMEBREW_CELLAR@@/tcl-tk/8.6.12_1/lib/libtk8.6.dylib
==> Changing install name in /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/bin/wish8.6
  from /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
    to @@HOMEBREW_CELLAR@@/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/itcl4.2.2/libitcl4.2.2.dylib
  from /opt/bb-brew/opt/tcl-tk/lib/itcl4.2.2/libitcl4.2.2.dylib
    to @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/itcl4.2.2/libitcl4.2.2.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/itk4.1.0/libitk4.1.0.dylib
  from /opt/bb-brew/opt/tcl-tk/lib/itk4.1.0/libitk4.1.0.dylib
    to @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/itk4.1.0/libitk4.1.0.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
  from /opt/bb-brew/opt/tcl-tk/lib/libtcl8.6.dylib
    to @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/libtcl8.6.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtk8.6.dylib
  from /opt/bb-brew/opt/tcl-tk/lib/libtk8.6.dylib
    to @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/libtk8.6.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/sqlite3.36.0/libsqlite3.36.0.dylib
  from /opt/bb-brew/opt/tcl-tk/lib/sqlite3.36.0/libsqlite3.36.0.dylib
    to @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/sqlite3.36.0/libsqlite3.36.0.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/tcllibc/macosx-x86_64-clang/tcllibc.dylib
  from /opt/bb-brew/opt/tcl-tk/lib/tcllibc/macosx-x86_64-clang/v3118_00000000000000000000000000000154.dylib
    to @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/tcllibc/macosx-x86_64-clang/v3118_00000000000000000000000000000154.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/tcltls1.7.22/tcltls.dylib
  from shared-tcltls.dylib
    to 
==> Changing install name in /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/bin/tclsh8.6
  from @@HOMEBREW_CELLAR@@/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
    to /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
==> Changing install name in /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/bin/wish8.6
  from @@HOMEBREW_CELLAR@@/tcl-tk/8.6.12_1/lib/libtk8.6.dylib
    to /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtk8.6.dylib
==> Changing install name in /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/bin/wish8.6
  from @@HOMEBREW_CELLAR@@/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
    to /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/itcl4.2.2/libitcl4.2.2.dylib
  from @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/itcl4.2.2/libitcl4.2.2.dylib
    to /opt/bb-brew/opt/tcl-tk/lib/itcl4.2.2/libitcl4.2.2.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/itk4.1.0/libitk4.1.0.dylib
  from @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/itk4.1.0/libitk4.1.0.dylib
    to /opt/bb-brew/opt/tcl-tk/lib/itk4.1.0/libitk4.1.0.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtcl8.6.dylib
  from @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/libtcl8.6.dylib
    to /opt/bb-brew/opt/tcl-tk/lib/libtcl8.6.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/libtk8.6.dylib
  from @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/libtk8.6.dylib
    to /opt/bb-brew/opt/tcl-tk/lib/libtk8.6.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/sqlite3.36.0/libsqlite3.36.0.dylib
  from @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/sqlite3.36.0/libsqlite3.36.0.dylib
    to /opt/bb-brew/opt/tcl-tk/lib/sqlite3.36.0/libsqlite3.36.0.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/tcllibc/macosx-x86_64-clang/tcllibc.dylib
  from @@HOMEBREW_PREFIX@@/opt/tcl-tk/lib/tcllibc/macosx-x86_64-clang/v3118_00000000000000000000000000000154.dylib
    to /opt/bb-brew/opt/tcl-tk/lib/tcllibc/macosx-x86_64-clang/v3118_00000000000000000000000000000154.dylib
==> Changing dylib ID of /opt/bb-brew/Cellar/tcl-tk/8.6.12_1/lib/tcltls1.7.22/tcltls.dylib
  from shared-tcltls.dylib
    to 
Error: new ID must be a String
/opt/bb-brew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/ruby-macho-3.0.0/lib/macho/macho_file.rb:318:in `change_dylib_id'
/opt/bb-brew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/ruby-macho-3.0.0/lib/macho/tools.rb:26:in `change_dylib_id'
/opt/bb-brew/Homebrew/Library/Homebrew/os/mac/keg.rb:10:in `change_dylib_id'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:26:in `block (2 levels) in relocate_dynamic_linkage'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/pathname.rb:341:in `ensure_writable'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:23:in `block in relocate_dynamic_linkage'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:22:in `each'
/opt/bb-brew/Homebrew/Library/Homebrew/extend/os/mac/keg_relocate.rb:22:in `relocate_dynamic_linkage'
/opt/bb-brew/Homebrew/Library/Homebrew/keg_relocate.rb:130:in `replace_placeholders_with_locations'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:501:in `block (2 levels) in bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/utils.rb:399:in `ignore_interrupts'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:499:in `ensure in block in bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:499:in `block in bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/keg.rb:334:in `block in lock'
/opt/bb-brew/Homebrew/Library/Homebrew/lock_file.rb:35:in `with_lock'
/opt/bb-brew/Homebrew/Library/Homebrew/keg.rb:330:in `lock'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:390:in `bottle_formula'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:105:in `block in bottle'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:104:in `each'
/opt/bb-brew/Homebrew/Library/Homebrew/dev-cmd/bottle.rb:104:in `bottle'
/opt/bb-brew/Homebrew/Library/Homebrew/brew.rb:110:in `<main>'

**Comment by carlocab** [comment]
I can reproduce this, but I'm still not quite sure why it's happening. Just to double-check: what is the output of `brew install --debug tcl-tk`?

**Comment by imresteiner** [comment]
[console_log.log](https://github.com/Homebrew/brew/files/7994998/console_log.log)

**Comment by carlocab** [comment]
My last theory was that this is a regression in `ruby-macho`, but I, unfortunately, have not been able to reproduce the problematic behaviour using just `ruby-macho`.

The problem is that `tcltls.dylib` has an install name of `shared-tcltls.dylib` (i.e. basename-only). Normally, after the install, [`fix_dynamic_linkage`](https://github.com/Homebrew/brew/blob/7b1e47558b5a544eb819b9af9b83db3e30386c8c/Library/Homebrew/extend/os/mac/keg_relocate.rb#L42) will turn this into an absolute path (i.e. something like `/usr/local/opt/tcl-tk/lib/tcltls1.7.22/tcltls.dylib`).

For some reason, `fix_dynamic_linkage` fails to do this, but doesn't complain about it. This only proves to be a problem when we try to bottle `tcl-tk`, and the call to [`change_dylib_id`](https://github.com/Homebrew/brew/blob/8a787e45ff45583fb245030757296548f82c1cbd/Library/Homebrew/extend/os/mac/keg_relocate.rb#L26) here chokes because the new ID is for some reason an empty string.

The right fix is probably to make sure `fix_dynamic_linkage` gets the dylib ID change right the first time (this is probably [`dylib_id_for`](https://github.com/Homebrew/brew/blob/7b1e47558b5a544eb819b9af9b83db3e30386c8c/Library/Homebrew/extend/os/mac/keg_relocate.rb#L100)).

But that still leaves open the question of how this even got broken in the first place. You could try reverting my commit from the [`git bisect` above](https://github.com/Homebrew/brew/issues/12832#issuecomment-1028864039), but I would be very surprised if it helped. (Maybe there's some edge case I'm not seeing.)

**Comment by Pencab** [comment]
CONTRIBUTING.md

**Comment by gromgit** [comment]
@carlocab, I can confirm that `grpc` correctly bottles after reverting https://github.com/Homebrew/brew/commit/568bc94f3ee66bb62a31ae9d9c7f15de8c60cc92. Here's the `brew install --build-bottle` and `stool` logs before reverting:

[grpc.build-bottle.v1.log](https://github.com/Homebrew/brew/files/8048254/grpc.build-bottle.v1.log)
[grpc.otool.v1.log](https://github.com/Homebrew/brew/files/8048257/grpc.otool.v1.log)

and after reverting:

[grpc.build-bottle.v2.log](https://github.com/Homebrew/brew/files/8048184/grpc.build-bottle.v2.log)
[grpc.otool.v2.log](https://github.com/Homebrew/brew/files/8048243/grpc.otool.v2.log)

Note that pre-revert, even though the log says:
```
==> Changing dylib ID of /opt/homebrew/Cellar/grpc/1.43.2/lib/libgrpc++_test_config.1.43.2.dylib
  from @rpath/libgrpc++_test_config.1.43.dylib
    to /opt/homebrew/opt/grpc/lib/libgrpc++_test_config.1.43.dylib
```
the change doesn't seem to be applied:
```
/opt/homebrew/opt/grpc/lib/libgrpc++_test_config.1.43.dylib:
	@rpath/libgrpc++_test_config.1.43.dylib (compatibility version 1.43.0, current version 1.43.2)
```

**Comment by carlocab** [comment]
One thing I am confused about: https://github.com/Homebrew/brew/commit/568bc94f3ee66bb62a31ae9d9c7f15de8c60cc92 only affects the rewriting of `RPATH`s, but the problem in `brew` is about the rewriting of dylib IDs. Why does the implementation of `RPATH` rewriting affect dylib ID rewriting?

**Comment by gromgit** [comment]
If I'm reading the commit correctly, it replaces an open-edit-write sequence on a dylib filename (`MachO::Tools.delete_rpath`) with an edit-write sequence on an already-open dylib (`Pathname.delete_rpath`). Perhaps there's a cache overwrite or some similar issue in the latter logic?

**Comment by carlocab** [comment]
I still don't see how that's related to our reading of dylib IDs (and subsequent re-reading and replacing) from the cache, though.

**Comment by gromgit** [comment]
I'm not sure either, and I'm definitely not familiar with the codebase, but since that's the only effective logic difference, that suggests a failure to commit existing changes somewhere, or overlapping changes.

Some years back, I wrote some code that ending up doing something like the following:
* read file F into buffer A
* make change in A
* read file F into buffer B
* make change B
* flush A to file
* flush B to file

Needless to say, the changes in A disappeared without a trace.

**Comment by carlocab** [comment]
This was surprisingly subtle. (Though maybe I should know better than to be surprised.)

Thanks for the help figuring this out, @gromgit, and thanks for helping track down the offending commit, @imresteiner.

Should be closed by #12864.
